design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,432 @@
+AWS, Azure and GCP is due.
+
+## Goals
+
+We have the following goals for providers:
+
+* The cloud vendors should maintain their Crossplane provider.
+* The users should have the ability to provision all of their infrastructure
+  through Crossplane, i.e. coverage.
+* The providers should be mature enough for users to depend on them in
+  production, i.e. maturity.
+
+This document aims for a decision for the following providers:
+* AWS
+  * Classic with [124 CRDs][provider-aws-crds] with 37 beta CRDs.
+  * Jet with [763 CRDs][jet-aws-crds], no beta CRDs.
+* Azure
+  * Classic with [21 CRDs][provider-azure-crds] with 7 beta CRDs.
+  * Jet with [647 CRDs][jet-azure-crds], no beta CRDs.
+* Google Cloud
+  * Classic with [27 CRDs][provider-gcp-crds] with 11 beta CRDs.
+  * Jet with [438 CRDs][jet-gcp-crds], no beta CRDs.
+
+Both variants of each cloud reach those goals at different levels but none of
+them are maintained by their cloud vendor. The aim of this document is to
+capture a decision about the future of these provider implementations with the
+goals stated above in mind.
+
+## Proposal
+
+There were three options in the [provider strategy doc][strategy-doc]:
+* Option A: Completely separate providers.
+  * The initial decision.
+* Option B: One provider but Jet-based resources in their own API group.
+* Option C: One provider and no Jet vs classic difference. Only API version
+  would be bumped if the implementation changes.
+
+We should go with the Option C, i.e. use Terrajet in
+[`provider-aws`][provider-aws], [`provider-azure`][provider-azure] and
+[`provider-gcp`][provider-gcp] as another way of generating CRDs and controllers
+so that they can reach the goal of coverage really fast. Then the community can
+keep investing in that single provider and make each CRD mature enough to depend
+on in production. In this way, we can increase the usage by increasing both the
+coverage and the maturity of the CRDs in a consolidated way. This increase in
+usage can attract cloud vendors to come and join the maintainership of these
+providers. Once they do, they can either progressively convert the
+implementation of the controllers to their choice of technology or start with
+new providers which we can suggest people to use and we can provide migration
+tooling from the community-maintained one. In either case, the users would see a
+clear path forward and we'd not waste any efforts by the community.
+
+Note that we are making two decisions here:
+* There will be a single provider.
+* The single provider will be the classic one and we'll add Terrajet CRDs on
+  top.
+
+The order of execution will be roughly as following, taking
+[`provider-aws`][provider-aws] as an example:
+1. Add Terrajet pipeline to the provider repository.
+1. Copy the manual resource configurations from provider-jet-aws and introduce
+  the CRDs of those configurations in their new API group, without changing
+  their version, i.e. `v1alpha2`.
```

Comment:
I expect some conflicts due to different resource grouping while merging the two. Should we clarify how to solve such conflicts in this document or discuss case by case basis?

For example, Elasticache Replication Group: [native](https://github.com/crossplane/provider-aws/blob/master/apis/cache/v1beta1/register.go#L28) under **cache** vs [jet](https://github.com/crossplane-contrib/provider-jet-aws/blob/release-0.4-preview/apis/elasticache/v1alpha2/zz_replicationgroup_types.go) under **elasticache** group.

Reply:
Just a brief note:

This decision tree might get more complicated for provider-aws. Original provider-aws CRDs that do not align with ack-generated code are currently kept under a `manualv1alpha1` apiVersion. (It looks like this primarily impacts ec2 and eks). So we would be doing a 3-way evaluation, there, for "manual code," "ack-generated," and "jet-generated."